#### Context: Explore US Bikeshare Data
 
You are assigned the task to analyze the Bicycle-sharing data for three major cities in the United States—Chicago, New York City, and Washington.
The datasets used for this project contain bike share data for the first six months of 2017. The csv files can be found in the same folder. The data is provided by Motivate, which is a bike share system provider for many cities in the United States. The data files for all three cities contain the same six columns:

●      Start Time (e.g., 2017-01-01 00:07:57)

●      End Time (e.g., 2017-01-01 00:20:53)

●      Trip Duration (in seconds - e.g., 776)

●      Start Station (e.g., Broadway & Barry Ave)

●      End Station (e.g., Sedgwick St & North Ave)

●      User Type (Subscriber or Customer)
 
The Chicago and New York City files also contain the following two columns:
    
●      Gender

●      Birth Year








In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
df_chi = pd.read_csv('data/RR-DPL-UA1 - chicago.csv')
df_nyc = pd.read_csv('data/RR-DPL-AU1 - new_york_city.csv')
df_wc = pd.read_csv('data/RR-DPL-AU1 - washington.csv')

In [3]:
df_chi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1551505 entries, 0 to 1551504
Data columns (total 8 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   Start Time     1551505 non-null  object 
 1   End Time       1551505 non-null  object 
 2   Trip Duration  1551505 non-null  int64  
 3   Start Station  1551505 non-null  object 
 4   End Station    1551505 non-null  object 
 5   User Type      1551505 non-null  object 
 6   Gender         1234638 non-null  object 
 7   Birth Year     1234822 non-null  float64
dtypes: float64(1), int64(1), object(6)
memory usage: 94.7+ MB


In [4]:
df_nyc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6816254 entries, 0 to 6816253
Data columns (total 8 columns):
 #   Column         Dtype  
---  ------         -----  
 0   Start Time     object 
 1   End Time       object 
 2   Trip Duration  int64  
 3   Start Station  object 
 4   End Station    object 
 5   User Type      object 
 6   Gender         object 
 7   Birth Year     float64
dtypes: float64(1), int64(1), object(6)
memory usage: 416.0+ MB


In [5]:
df_wc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1751452 entries, 0 to 1751451
Data columns (total 6 columns):
 #   Column         Dtype  
---  ------         -----  
 0   Start Time     object 
 1   End Time       object 
 2   Trip Duration  float64
 3   Start Station  object 
 4   End Station    object 
 5   User Type      object 
dtypes: float64(1), object(5)
memory usage: 80.2+ MB


In [6]:
# Combine 3 dataframes with same column names using concat()
df = pd.concat([df_chi, df_nyc, df_wc], axis=0, join='inner')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10119211 entries, 0 to 1751451
Data columns (total 6 columns):
 #   Column         Dtype  
---  ------         -----  
 0   Start Time     object 
 1   End Time       object 
 2   Trip Duration  float64
 3   Start Station  object 
 4   End Station    object 
 5   User Type      object 
dtypes: float64(1), object(5)
memory usage: 540.4+ MB


In [7]:
# verify combined dataframe contains all data from 3 dataframes
print((df_chi.shape[0]+df_nyc.shape[0]+df_wc.shape[0])==(df.shape[0]))

True


In [8]:
# Convert the date column to datetime64
df['s_time'] = pd.to_datetime(df['Start Time'])
df['e_time'] = pd.to_datetime(df['End Time'])

# Set s_time column as index
df.set_index('s_time',inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 10119211 entries, 2017-01-01 00:00:36 to 2017-06-30 12:59:59
Data columns (total 7 columns):
 #   Column         Dtype         
---  ------         -----         
 0   Start Time     object        
 1   End Time       object        
 2   Trip Duration  float64       
 3   Start Station  object        
 4   End Station    object        
 5   User Type      object        
 6   e_time         datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(5)
memory usage: 617.6+ MB


In [10]:
df = df.drop(columns = ['Start Time', 'End Time'])
df.head(3)

Trip Duration                         Start Station  \
s_time                                                                     
2017-01-01 00:00:36          356.0                  Canal St & Taylor St   
2017-01-01 00:02:54          327.0            Larrabee St & Menomonee St   
2017-01-01 00:06:06          745.0  Orleans St & Chestnut St (NEXT Apts)   

                                      End Station   User Type  \
s_time                                                          
2017-01-01 00:00:36      Canal St & Monroe St (*)    Customer   
2017-01-01 00:02:54  Sheffield Ave & Kingsbury St  Subscriber   
2017-01-01 00:06:06    Ashland Ave & Blackhawk St  Subscriber   

                                 e_time  
s_time                                   
2017-01-01 00:00:36 2017-01-01 00:06:32  
2017-01-01 00:02:54 2017-01-01 00:08:21  
2017-01-01 00:06:06 2017-01-01 00:18:31

#### Task1: What is the most popular month for start time?

In [11]:
df_month = df.groupby(df.index.month).size().sort_values(ascending=False)
df_month

s_time
6    2635721
5    2209083
4    1950220
2    1184293
3    1126472
1    1013422
dtype: int64

#### The most popular month is June.

#### Task2: What is the most popular day of week (Monday, Tuesday, etc.) for start time?

In [30]:
df_day_name = df.groupby(df.index.day_name()).size().sort_values(ascending=False)  # .dayofweek .day_name() 有括号因为它有参数改变语言，默认英语
df_day_name.head(3)

s_time
Wednesday    1680186
Thursday     1564066
Tuesday      1493367
dtype: int64

#### The most popular day of week is Wednesday.

#### Task3: What is the most popular hour of day for start time?

In [13]:
df_hour = df.groupby(df.index.hour).size().sort_values(ascending=False)
df_hour.head(3)

s_time
17    948086
8     876159
18    832063
dtype: int64

#### The most popular hour of day is 17:00.

#### Task4: What is the total trip duration and average trip duration?

In [14]:
total_duration = df['Trip Duration'].sum()
print(f'Total trip duration = {round(total_duration)} seconds')

Total trip duration = 10213563095 seconds


In [15]:
avg_duration = df['Trip Duration'].mean()
print(f'Average trip duration = {round(avg_duration)} seconds')

Average trip duration = 1009 seconds


#### Task5: What is the most popular start station and most popular end station?

In [16]:
df_s_station = df['Start Station'].groupby(by=df['Start Station']).count().sort_values(ascending=False)
df_s_station

Start Station
Pershing Square North             71108
E 17 St & Broadway                48311
Broadway & E 22 St                48044
W 21 St & 6 Ave                   46068
West St & Chambers St             43716
                                  ...  
Ashland Ave & 66th St                 2
NYCBS Depot - STY - Valet Scan        2
May St & 69th St                      2
State St & 76th St                    2
Ashland Ave & Garfield Blvd           1
Name: Start Station, Length: 1713, dtype: int64

In [17]:
print(f"The most popular start station is {df_s_station.index[0]}.")

The most popular start station is Pershing Square North.


#### The most popular start station is Pershing Square North.

In [18]:
df_e_station = df['End Station'].groupby(by=df['End Station']).count().sort_values(ascending=False)
df_e_station

End Station
Pershing Square North     70048
Broadway & E 22 St        52409
E 17 St & Broadway        51587
W 21 St & 6 Ave           46686
West St & Chambers St     45825
                          ...  
Phillips Ave & 79th St        2
State St & 76th St            1
JC Medical Center             1
E 15 St & Irving Pl           1
Jersey & 3rd                  1
Name: End Station, Length: 1731, dtype: int64

In [19]:
print(f"The most popular end station is {df_e_station.index[0]}.")

The most popular end station is Pershing Square North.


#### The most popular end station is Pershing Square North.

#### Task6: What is the most popular trip?

In [20]:
df_trip = df.groupby(['Start Station','End Station']).size().sort_values(ascending=False)
df_trip

Start Station                                          End Station                                          
Lake Shore Dr & Monroe St                              Streeter Dr & Grand Ave                                  4647
Jefferson Dr & 14th St SW                              Jefferson Dr & 14th St SW                                4047
Streeter Dr & Grand Ave                                Streeter Dr & Grand Ave                                  3853
E 7 St & Avenue A                                      Cooper Square & E 7 St                                   3641
Smithsonian-National Mall / Jefferson Dr & 12th St SW  Smithsonian-National Mall / Jefferson Dr & 12th St SW    3370
                                                                                                                ... 
Graham Ave & Herbert St                                8 Ave & W 16 St                                             1
                                                       9 St & 44 Rd     

In [21]:
# transform index into normal columns
df_trip = df_trip.reset_index()

In [22]:
print(f"The most popular trip is start at {df_trip.iloc[0,1]} and end at {df_trip.iloc[0,2]}.")

The most popular trip is start at Streeter Dr & Grand Ave and end at 4647.


#### The most popular trip is start at Streeter Dr & Grand Ave and end at 4647.

#### Task7: What are the counts of each user type?


In [23]:
df_user_type = df.groupby(df['User Type']).size()

In [24]:
print(f"counts of Customer is {df_user_type['Customer']}.")
print(f"counts of Dependent is {df_user_type['Dependent']}.")
print(f"counts of Subscriber is {df_user_type['Subscriber']}.")

counts of Customer is 1463607.
counts of Dependent is 4.
counts of Subscriber is 8639691.
